scipy/linalg/_decomp_update.pyx

```diff
@@ -0,0 +1,2416 @@
+    sigma_min = wnorm / sigma_max
+    rc = sigma_min / sigma_max
+
+    # check the conditioning of the problem.
+    if blas_t_less_than(rc, RCOND[0]):
+        RCOND[0] = rc
+        return 2
+    RCOND[0] = rc
+
+    if blas_t_less_than(inv_root2, wnorm):
+        scal(m, 1/wnorm, u, us[0])
+        scal(n, unorm, s, 1)
+        s[n] = unorm*wnorm
+        return 0
+
+    # if the above check failed, try one reorthogonalization
+    if qisF:
+        if blas_t is float or blas_t is double:
+            gemv(T, m, n, 1, q, qs[1], u, us[0], 0, s+n, 1)
+        else:
+            gemv(C, m, n, 1, q, qs[1], u, us[0], 0, s+n, 1)
+        gemv(N, m, n, -1, q, qs[1], s+n, 1, 1, u, us[0])
+    else:
+        if blas_t is float or blas_t is double:
+            gemv(N, n, m, 1, q, n, u, us[0], 0, s+n, 1)
+        else:
+            blas_t_conj(m, u, us)
+            gemv(N, n, m, 1, q, n, u, us[0], 0, s+n, 1)
+            blas_t_conj(m, u, us)
+            blas_t_conj(n, s+n, &ss)
+        gemv(T, n, m, -1, q, n, s+n, 1, 1, u, us[0])
+        
+    wpnorm = nrm2(m, u, us[0]) 
+    
+    if blas_t_less_than(wpnorm, wnorm/inv_root2): # u lies in span(q) 
+        scal(m, 0, u, us[0])
+        axpy(n, 1, s, 1, s+n, 1)
+        scal(n, unorm, s, 1)
+        s[n] = 0
+        return 1
+    scal(m, 1/wpnorm, u, us[0])
+    axpy(n, 1, s, 1, s+n, 1)
+    scal(n, unorm, s, 1)
+    s[n] = wpnorm*unorm
+    return 0
+
+def _form_qTu(object a, object b):
+    """ this function only exists to expose the cdef version below for testing
+        purposes. Here we perform minimal input validation to ensure that the
+        inputs meet the requirements below.
+    """
+    cdef cnp.ndarray q, u, qTu
+    cdef int typecode
+    cdef void* qTuvoid
+    cdef int qTus[2]
+
+    if not cnp.PyArray_Check(a) or not cnp.PyArray_Check(b):
+        raise ValueError('Inputs must be arrays')
+
+    q = a
+    u = b
+    
+    typecode = cnp.PyArray_TYPE(q)
+    if cnp.PyArray_TYPE(u) != typecode:
+        raise ValueError('q and u must have the same type.')
+
+    if not (typecode == cnp.NPY_FLOAT or typecode == cnp.NPY_DOUBLE \
+            or typecode == cnp.NPY_CFLOAT or typecode == cnp.NPY_CDOUBLE):
+        raise ValueError('q and u must be a blas compatible type: f d F or D')
+
+    q = validate_array(q, True)
+    u = validate_array(u, True)
+
+    qTu = cnp.PyArray_ZEROS(u.ndim, u.shape, typecode, 1)
+    qTuvoid = extract(qTu, qTus)
+    form_qTu(q, u, qTuvoid, qTus, 0)
+    return qTu
+
+cdef form_qTu(cnp.ndarray q, cnp.ndarray u, void* qTuvoid, int* qTus,
+              int k):
+    """ assuming here that q and u have compatible shapes, and are the same type
+        + Q is contiguous.  This function is preferable over simply calling
+        np.dot for two reasons: 1) this output is always in F order, 2) no
+        copies need be made if Q is complex.  Point 2 in particular makes this
+        a good bit faster for complex inputs.
```

Comment:
if you still remember the timings, might be good to put an approximate factor here. "a good bit" could be 20% or a factor of 5